pkg/squash/vfs_fs.go

```diff
@@ -0,0 +1,210 @@
+		return nil, fs.ErrInvalid
+	}
+	lnk, err := fsGet(f, name, false, false)
+	if err != nil {
+		return nil, err
+	}
+	tgt, err := lnk.Get(".", false, true)
+	if err != nil {
+		return nil, err
+	}
+	return &fsfileReader{ //nolint:exhaustivestruct
+		origName: name,
+		tgt:      tgt,
+		lnk:      lnk,
+	}, nil
+}
+
+type fsfileReader struct {
+	origName string
+	tgt      *fsfile
+	lnk      *fsfile
+
+	mu      sync.Mutex
+	pos     int
+	closed  bool
+	dirents []fs.DirEntry
+}
+
+// Open implements io/fs.File
+func (f *fsfileReader) Stat() (fs.FileInfo, error) {
+	if f.tgt.header == nil {
+		return nil, &fs.PathError{
+			Op:   "stat",
+			Path: f.origName,
+			Err:  ErrMissing,
+		}
+	}
+	hdr := *f.tgt.header // shallow copy
+	hdr.Name = f.lnk.header.Name
+	return hdr.FileInfo(), nil
+}
+
+// Open implements io/fs.File
+func (f *fsfileReader) Read(buf []byte) (_ int, err error) {
+	defer func() {
+		if err != nil && !errors.Is(err, io.EOF) {
+			err = &fs.PathError{
+				Op:   "read",
+				Path: f.origName,
+				Err:  err,
+			}
+		}
+	}()
+	if f.tgt.header == nil || f.tgt.header.Typeflag == tar.TypeDir {
+		return 0, ErrIsDir
+	}
+	if int64(len(f.tgt.body)) < f.tgt.header.Size {
+		return 0, ErrMissing
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.closed {
+		return 0, fs.ErrClosed
+	}
+	if int64(len(f.tgt.body)) < f.tgt.header.Size {
+		return 0, ErrMissing
+	}
+	if f.pos == len(f.tgt.body) {
+		return 0, io.EOF
+	}
+	n := copy(buf, f.tgt.body[f.pos:])
+	f.pos += n
+	return n, nil
+}
+
+// Open implements io/fs.File
+func (f *fsfileReader) Close() error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.closed {
+		return fs.ErrClosed
+	}
+	f.closed = true
+	return nil
+}
+
+type missingDirEntry string
+
+func (name missingDirEntry) Name() string          { return string(name) }
+func (missingDirEntry) IsDir() bool                { return true }
+func (missingDirEntry) Type() fs.FileMode          { return fs.ModeDir }
+func (missingDirEntry) Info() (fs.FileInfo, error) { return nil, ErrMissing }
+
+func (name missingDirEntry) String() string { return fmt.Sprintf("<missingDirEntry %q>", string(name)) }
+
+type dirEntry struct {
+	tgt fs.DirEntry
+	lnk fs.DirEntry
+}
+
+func (d *dirEntry) Name() string               { return d.lnk.Name() }
+func (d *dirEntry) IsDir() bool                { return d.lnk.IsDir() }
+func (d *dirEntry) Type() fs.FileMode          { return d.lnk.Type() }
+func (d *dirEntry) Info() (fs.FileInfo, error) { return d.lnk.Info() }
+
+func (d *dirEntry) String() string { return fmt.Sprintf("squash.dirEntry{%q}", d.Name()) }
+
+// Open implements io/fs.ReadDirFile
+func (f *fsfileReader) ReadDir(n int) ([]fs.DirEntry, error) { //nolint:varnamelen // obvious
+	if f.tgt.header != nil && f.tgt.header.Typeflag != tar.TypeDir {
+		return nil, ErrNotDir
+	}
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	if f.closed {
+		return nil, fs.ErrClosed
+	}
+	if f.dirents == nil {
+		f.dirents = make([]fs.DirEntry, 0, len(f.tgt.children))
```

Comment:
Whoa, I was not expecting `f.dirents` to be lazily cached here, for some reason. Comments!! 🙂

Reply:
I've added a comment to the struct definition